NIFI-1420 Adding Splunk bundle

bbende opened the pull request:
## Overview
This pull request adds a Splunk bundle with the following processors:
 
* **ListenSplunkForwarder** - Receives data from a Splunk forwarder, based off the AbstractListenEventProcessor created during RELP work.

* **PutSplunk** - Delivers to Splunk Enterprise over TCP or UDP, based some of the design off PutKafka for how to handle delimited messages, also created an AbstractPutEventProcessor to  make future "Put" TCP/UDP processors easier, involved refactoring code from PutSyslog.

* **GetSplunk** - Extracts data from Splunk Enterprise based on a query, processor provides ability to optionally specify a time range, or have the processor manage the time ranges using the new state management API.

## Testing

### Download and extract the forwarder and Splunk enterprise:
http://www.splunk.com/en_us/download/universal-forwarder.html
http://www.splunk.com/en_us/download/splunk-enterprise.html

## Splunk Forwarder:

Edit/Create splunkforwarder/etc/system/local/outputs.conf and configure an output for NiFi to listen to:

    [tcpout:nifi]
    server=localhost:6588
    sendCookedData=false

Start the forwarder:

    ./splunkforwarder/bin/splunk start

At this point you can create a NiFi flow with ListenSplunkForwarder, listening on TCP port 6588 and should be receiving data.

## Splunk Enterprise

Start Splunk Enterprise

    ./splunk/bin/splunk start

Go to http://localhost:8000 in your browser

Create an input from the Settings -> Data Inputs menu, create a TCP or UDP input.

After that you should be able to use PutSplunk to deliver data to the input created in the previous step, and GetSplunk to extract data.

olegz commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/put/AbstractPutEventProcessor.java:
```diff
@@ -0,0 +1,474 @@
+        relationships.add(REL_FAILURE);
+        relationships.addAll(getAdditionalRelationships());
+        this.relationships = Collections.unmodifiableSet(relationships);
+    }
```
I know it's a common pattern  used by processors, but given that _init()_ method is actually executed multiple times during the life-cycle of the Processor, the above approach would essentially be re-creating the same collections (an the GC) every time _init()_ is called. So consider static initializer.

olegz commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/put/AbstractPutEventProcessor.java:
```diff
@@ -0,0 +1,474 @@
+     */
+    protected List<PropertyDescriptor> getAdditionalProperties() {
+        return Collections.EMPTY_LIST;
+    }
```
Is it really necessary to override if that's a default behavior?

bbende replied:
The idea was to make getSupportedPropertyDescriptors() final so that the properties provided by the base class had to be there, and then provide getAdditionalProperties() to allow a subclass to add additional properties.

olegz commented on nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/put/sender/ChannelSender.java:
```diff
@@ -0,0 +1,103 @@
+/**
+ * Base class for sending messages over a channel.
+ */
+public abstract class ChannelSender {
```
Just a general comment. Unless you already know that this class should be public, consider reducing visibility. We can always make it public, but we can never go the other direction. Not sure if it applies here. . . your call.

bbende replied:
Agreed, will consider that.

olegz commented on nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/main/java/org/apache/nifi/processors/splunk/ListenSplunkForwarder.java:
```diff
@@ -0,0 +1,208 @@
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        final int maxBatchSize = context.getProperty(MAX_BATCH_SIZE).asInteger();
```
Considering that this is using the similar batching approach used in PutKafka, curious if you tested it with run-duration > 0. See comments in https://issues.apache.org/jira/browse/NIFI-1534 as it may be suffering from the same issue.

bbende replied:
I had submitted this before the PutKafka issue was discovered, so it is possible that there could be a similar problem. However, from reading the commentary on NIFI-1534, it sounds like the issue only occurs when @SupportsBatching is used, and PutSplunk does not have that annotation.

trixpan commented:
@bbende

Out of curiosity, why would someone use ListenSplunk instead of ListenSyslog using TCP with Parsing disabled (aka poor man ListenTCP) ?

I was under the impression Splunk forwarder frame format when using non indexed data is a pure TCP stream, pretty much like netcat.  (Or are we going to be able to acknowledge events as the Indexer does when communicating with forwarders?)

If that is the case, shouldn't we call this ListenTCP instead of ListenSplunkForwarder?

trixpan commented:
@bbende 

If the issue is the separator, maybe we should then add a drop down "flavour" selection with well known separator combinations ( '\n\n', '\n', etc)

bbende commented:
@trixpan You are right that ListenSplunkForwarder is essentially a ListenTCP. I started out thinking we might be able to do acknowledgements, but was unable to find any documentation on the acknowledgement protocol and what the frames look like, so in its current state it is just a straight TCP listener. If you know how the acknowledgements work, or know of a resource that describes it, please let me know.

Part of the issue comes down to user perception... since we are creating a splunk bundle and going to have PutSplunk and GetSplunk, a potential user might see that and come to the conclusion that we don't have a Listener for Splunk, since ListenTCP would be part of standard processors.

I think for the future we should definitely have a ListenTCP that is as generic as possible and exposes options like you suggested about the line delimiter, and other ones about how to identify messages based on patters or length.

trixpan commented:
@bbende 

The protocol is proprietary

https://answers.splunk.com/answers/12543/splunk-data-format.html

  so unless they suddenly decided to document it the ListenSplunk will continue be a ListenTCP (hence my previous comments). 

I am quite picky on this one as certain Splunk SE teams seem to have this culture of demotivating the use of  TCP forwarding as an "unfair"  use of their proprietary IP (technically speaking -  legal is gray area -  you can in fact use splunk forwarder as a "miNiFi"-like agent and as long you don't send data into tge indexers your licence will not be consumed).

JPercivall commented:
I observed a couple of interesting interactions when testing the PutSplunk processor. I'm not sure whether it is a problem with the processor, the Splunk Java SDK, Splunk Enterprise or my Splunk configurations:
* I set up the PutSplunk processor to communicate via UDP and created a UDP data input using the UI for Splunk Enterprise. The PutSplunk is transmitting messages at ~1 per second. While it's transmitting no new events are created yet in Splunk Enterprise (verified by viewing a real time view of the past 30 seconds). I stop the PutSplunk processor and then run a search in Splunk UI to see if any events came in on that port. There is one event that is registered at when I started the processor and contains effectively all (didn't count the seconds I was sending) the data that was generated. I'm not sure why all the UDP messages are grouped together. When I set PutSplunk to send messages every 20 seconds the UDP messages are treated as their own event.
* I set up PutSplunk to communicate via TCP using SSL. I looked for options to create a Data Input in Splunk Enterprise using TCP and SSL but couldn't find anything definitive (and all my searches turned up results for the Forwarder). I enabled SSL in Splunk Web by going to the general settings and notice that SSL properties are in the inputs.conf and server.conf files. I attempt to send data using PutSplunk using SSL to the TCP data input. The event is received and I see it in the Splunk UI in real time. The only problem being that it's still encoded (see image). Typically when I get errors with SSL there is some obscure error relating to truncation attacks or Cipher Suites but I didn't seem to get any errors.
![screen shot 2016-03-01 at 5 21 54 pm](https://cloud.githubusercontent.com/assets/11302527/13444066/2357d1e2-dfd2-11e5-8150-9684b7299a6d.png)

JPercivall commented:
Whenever an SSL context is being created (done 4 times in multiple different classes) "SSLContextService.ClientAuth.REQUIRED" is used. Should we offer configuration options for this? For PutSplunk and GetSplunk I can understand it because we are requiring the server to provide authorization but maybe someone using ListenSplunkForwader doesn't need it to verify who's sending the data.

bbende commented:
@JPercivall I pushed up three commits, one for each processor, that I believe address all your comments/suggestions. Let me know if anything is missing or not addressed.

Regarding the UDP issue you saw, this only appears to be an issue when selecting the source type as misc_text and I believe this is something more on the side of sending the appropriate data to appropriate index type. When I send syslog messages to a syslog type over UDP it works fine.

Regarding the SSL issue you saw, I can't explain how Splunk was able to accept the encrypted data in your case, but I did create a tcp-ssl input type in the local inputs.conf and was able to send data to it and have it show up normal in Splunk, so I think that is the intended use case. I don't see a way to make TCP inputs with SSL through the UI.

JPercivall commented:
Should have made this comment on the first commit but can the LogGenerator be put into a util package on the same path? It makes it more readable to only have the test classes in package "org.apache.nifi.processors.splunk".

bbende commented:
@JPercivall Pushed up another commit that addresses the additional comments from today.

Part of this change I decided to go the route that @trixpan suggested and change ListenSplunkForwarder to ListenTCP, and as a result moved it to the standard bundle. This will open it up to a lot more use cases and it wasn't really Splunk specific. As a result I decided to take out the mime.type attribute since it is writing bytes to FlowFiles and may not really be text/plain all the time.

Let me know if anything else needs updating or was left out.

bbende commented:
@JPercivall Pushed up a commit that addresses your comment about Client Auth being required, and I think I also fixed the race condition on the RELP test. If you are good with everything let me know and I will squash the commits.

JPercivall commented:
Just finished looking over the code and doing a round of tests. Everything looks good.

+1, Thanks for the contribution Bryan, looks like a very powerful new addition
